### ISSUE TITLE
bindings_with_variant_name is unhelpful and incorrect for function arguments

### ISSUE THREAD

**Issue by RDambrosio016** [rust-lang/rust#88730]
Given the following code:

```rust
#![allow(non_camel_case_types, unused_variables, dead_code)]

enum Foo {
    c
}

fn foo(c: Foo) {}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e6829290d29151853fdd98844ba6391c

The current output is:

```
warning[E0170]: pattern binding `c` is named the same as one of the variants of the type `Foo`
 --> src/lib.rs:7:8
  |
7 | fn foo(c: Foo) {}
  |        ^ help: to match on the variant, qualify the path: `Foo::c`
  |
  = note: `#[warn(bindings_with_variant_name)]` on by default
```

Ideally the lint should not trigger on such a pattern, it does not really make sense and the suggestion is wrong altogether,

**Comment by nbdd0121** [comment]
Note that the suggestion isn't wrong in this particular case, because `Foo::c` can actually match any Foo. It's wrong only when `Foo` has 2+ variants.

**Comment by RDambrosio016** [comment]
Right, the real world case of this was with the ascii crate with the AsciiChar enum, this is just a tiny test case
